README.md
poslv-ros
=============================

ROS wrapper around libposlv.


changelog
poslv-ros (0.1-44) precise; urgency=low

  * cleaning the code 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Fri, 26 Sep 2014 10:47:08 +0200

poslv-ros (0.1-43) trusty; urgency=low

  * porting to indigo

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Tue, 16 Sep 2014 18:52:01 +0200

poslv-ros (0.1-42) precise; urgency=low

  * issues with groovy 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Thu, 27 Mar 2014 13:41:39 +0100

poslv-ros (0.1-41) precise; urgency=low

  * adding pkg-config 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Tue, 25 Mar 2014 17:03:46 +0100

poslv-ros (0.1-40) precise; urgency=low

  * update version

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Sun, 14 Jul 2013 15:52:56 +0000

poslv-ros (0.1-39) precise; urgency=low

  * adding a sleep in service 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Tue, 11 Jun 2013 21:10:30 +0200

poslv-ros (0.1-38) precise; urgency=low

  * error in the service response 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Tue, 11 Jun 2013 21:01:56 +0200

poslv-ros (0.1-37) precise; urgency=low

  * adding service for switching rtcm to cmr 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Tue, 11 Jun 2013 12:33:52 +0000

poslv-ros (0.1-36) precise; urgency=low

  * security check for error codes 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Sun, 09 Jun 2013 15:56:22 +0000

poslv-ros (0.1-35) lucid; urgency=low

  * more debugging

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Fri, 07 Jun 2013 00:22:49 +0200

poslv-ros (0.1-34) lucid; urgency=low

  * more debugging output 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Thu, 06 Jun 2013 23:56:40 +0200

poslv-ros (0.1-33) precise; urgency=low

  * groovy distribution 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Thu, 06 Jun 2013 08:50:16 +0000

poslv-ros (0.1-32) precise; urgency=low

  * changing frame_id 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Wed, 05 Jun 2013 17:34:31 +0000

poslv-ros (0.1-31) precise; urgency=low

  * bug discovered on field, missing exception handler
    frequency for DMI too high 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Mon, 03 Jun 2013 11:50:26 +0000

poslv-ros (0.1-30) precise; urgency=low

  * adding check of inter packet time 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Fri, 31 May 2013 16:18:36 +0000

poslv-ros (0.1-29) precise; urgency=low

  * changing message members name 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Fri, 31 May 2013 15:21:07 +0000

poslv-ros (0.1-28) precise; urgency=low

  * make the node respawn 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Thu, 30 May 2013 17:55:50 +0000

poslv-ros (0.1-27) precise; urgency=low

  * update of remake for mixing ROS distributions 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Thu, 30 May 2013 13:46:42 +0000

poslv-ros (0.1-26) precise; urgency=low

  * removing nullptr

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Wed, 29 May 2013 14:44:08 +0000

poslv-ros (0.1-25) lucid; urgency=low

  * missing dependencies 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Wed, 29 May 2013 09:54:08 +0200

poslv-ros (0.1-24) precise; urgency=low

  * distribution on PPA 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Tue, 28 May 2013 18:40:06 +0000

poslv-ros (0.1-23) precise; urgency=low

  * catching object factory exception

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Tue, 28 May 2013 17:20:51 +0000

poslv-ros (0.1-22) precise; urgency=low

  * changing library name 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Tue, 28 May 2013 09:14:50 +0000

poslv-ros (0.1-21) precise; urgency=low

  * update frequency, packet counters, diagnostics

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Sat, 25 May 2013 10:47:10 +0000

poslv-ros (0.1-20) precise; urgency=low

  * change hardware ID for diagnostics 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Fri, 24 May 2013 10:19:41 +0000

poslv-ros (0.1-19) precise; urgency=low

  * adding secondary GPS status 

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Tue, 21 May 2013 14:04:50 +0000

poslv-ros (0.1-18) precise; urgency=low

  * porting to remake (corrected changelog)

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Tue, 14 May 2013 20:44:39 +0000

poslv-ros (0.1-16) precise; urgency=low

  * adding diagnostics
  * restructuring

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Tue, 14 May 2013 20:27:53 +0000

poslv-ros (0.1-15) precise; urgency=low

  * adding dependency

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Tue, 14 May 2013 20:27:53 +0000

poslv-ros (0.1-14) precise; urgency=low

  * final yaml

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Tue, 14 May 2013 20:27:53 +0000

poslv-ros (0.1-13) precise; urgency=low

  * testing

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Tue, 14 May 2013 20:27:53 +0000

poslv-ros (0.1-12) precise; urgency=low

  * same fiddling

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Tue, 14 May 2013 20:27:53 +0000

poslv-ros (0.1-11) precise; urgency=low

  * fiddling around with yaml

 -- Jerome Maye <jerome.maye@mavt.ethz.ch>  Tue, 14 May 2013 20:27:53 +0000

poslv-ros (0.1-10) precise; urgency=low

  * correcting yaml file


src/poslv/lib/PosLvNode.cpp
/******************************************************************************
 * Copyright (C) 2013 by Jerome Maye                                          *
 * jerome.maye@gmail.com                                                      *
 *                                                                            *
 * This program is free software; you can redistribute it and/or modify       *
 * it under the terms of the Lesser GNU General Public License as published by*
 * the Free Software Foundation; either version 3 of the License, or          *
 * (at your option) any later version.                                        *
 *                                                                            *
 * This program is distributed in the hope that it will be useful,            *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of             *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the              *
 * Lesser GNU General Public License for more details.                        *
 *                                                                            *
 * You should have received a copy of the Lesser GNU General Public License   *
 * along with this program. If not, see <http://www.gnu.org/licenses/>.       *
 ******************************************************************************/

#include "PosLvNode.h"

#include <bitset>

#include <diagnostic_updater/publisher.h>

#include <boost/shared_ptr.hpp>
#include <boost/make_shared.hpp>

#include <libposlv/types/VehicleNavigationSolution.h>
#include <libposlv/types/VehicleNavigationPerformance.h>
#include <libposlv/types/TimeTaggedDMIData.h>
#include <libposlv/types/PrimaryGPSStatus.h>
#include <libposlv/types/SecondaryGPSStatus.h>
#include <libposlv/types/GAMSSolutionStatus.h>
#include <libposlv/types/IINSolutionStatus.h>
#include <libposlv/types/GeneralStatusFDIR.h>
#include <libposlv/types/ProgramControl.h>
#include <libposlv/types/BaseGPS1Setup.h>
#include <libposlv/com/TCPConnectionClient.h>
#include <libposlv/sensor/POSLVComTCP.h>
#include <libposlv/exceptions/IOException.h>
#include <libposlv/exceptions/SystemException.h>
#include <libposlv/exceptions/TypeCreationException.h>
#include <libposlv/base/Timer.h>
#include <libposlv/base/Factory.h>
#include <libposlv/types/Packet.h>
#include <libposlv/types/Group.h>

#include "poslv/VehicleNavigationSolutionMsg.h"
#include "poslv/VehicleNavigationPerformanceMsg.h"
#include "poslv/TimeTaggedDMIDataMsg.h"

namespace poslv {

/******************************************************************************/
/* Constructors and Destructor                                                */
/******************************************************************************/

  PosLvNode::PosLvNode(const ros::NodeHandle& nh) :
      _nodeHandle(nh),
      _alignStatus(8),
      _navStatus1(-1),
      _navStatus2(-1),
      _vnsPacketCounter(0),
      _vnpPacketCounter(0),
      _dmiPacketCounter(0),
      _lastVnsTimestamp(0),
      _lastInterVnsTime(0),
      _lastVnpTimestamp(0),
      _lastInterVnpTime(0),
      _lastDmiTimestamp(0),
      _lastInterDmiTime(0),
      _gamsStatus(7),
      _iinStatus(8),
      _generalStatusA(0),
      _generalStatusB(0),
      _generalStatusC(0),
      _fdirLevel1Status(0),
      _fdirLevel2Status(0),
      _fdirLevel4Status(0),
      _fdirLevel5Status(0),
      _cmr0Count(0),
      _cmr1Count(0),
      _cmr2Count(0),
      _cmr94Count(0),
      _rtcm1Count(0),
      _rtcm3Count(0),
      _rtcm9Count(0),
      _rtcm18Count(0),
      _rtcm19Count(0) {
    _gpsStatusMsgs[-1] = "Unknown";
    _gpsStatusMsgs[0] = "No data from receiver";
    _gpsStatusMsgs[1] = "Horizontal C/A mode";
    _gpsStatusMsgs[2] = "3-dimension C/A mode";
    _gpsStatusMsgs[3] = "Horizontal DGPS mode";
    _gpsStatusMsgs[4] = "3-dimension DGPS mode";
    _gpsStatusMsgs[5] = "Float RTK mode";
    _gpsStatusMsgs[6] = "Integer wide lane RTK mode";
    _gpsStatusMsgs[7] = "Integer narrow lane RTK mode";
    _gpsStatusMsgs[8] = "P-code";
    _alignStatusMsgs[0] = "Full navigation";
    _alignStatusMsgs[1] = "Fine alignment active";
    _alignStatusMsgs[2] = "GC CHI 2";
    _alignStatusMsgs[3] = "PC CHI 2";
    _alignStatusMsgs[4] = "GC CHI 1";
    _alignStatusMsgs[5] = "PC CHI 1";
    _alignStatusMsgs[6] = "Coarse leveling active";
    _alignStatusMsgs[7] = "Initial solution assigned";
    _alignStatusMsgs[8] = "No valid solution";
    _gamsStatusMsgs[0] = "Fixed integer";
    _gamsStatusMsgs[1] = "Fixed integer test install data";
    _gamsStatusMsgs[2] = "Degraded fixed integer";
    _gamsStatusMsgs[3] = "Floated ambiguity";
    _gamsStatusMsgs[4] = "Degraded floated ambiguity";
    _gamsStatusMsgs[5] = "Solution without install data";
    _gamsStatusMsgs[6] = "Solution from navigator attitude and install data";
    _gamsStatusMsgs[7] = "No solution";
    _iinStatusMsgs[1] = "Fixed narrow lane RTK";
    _iinStatusMsgs[2] = "Fixed wide lane RTK";
    _iinStatusMsgs[3] = "Float RTK";
    _iinStatusMsgs[4] = "Code DGPS";
    _iinStatusMsgs[5] = "RTCM DGPS";
    _iinStatusMsgs[6] = "Autonmous (C/A)";
    _iinStatusMsgs[7] = "GPS navigation solution";
    _iinStatusMsgs[8] = "No solution";
    getParameters();
    _vehicleNavigationSolutionPublisher =
      _nodeHandle.advertise<poslv::VehicleNavigationSolutionMsg>(
      "vehicle_navigation_solution", _queueDepth);
    _vehicleNavigationPerformancePublisher =
      _nodeHandle.advertise<poslv::VehicleNavigationPerformanceMsg>(
      "vehicle_navigation_performance", _queueDepth);
    _timeTaggedDMIDataPublisher =
      _nodeHandle.advertise<poslv::TimeTaggedDMIDataMsg>(
      "time_tagged_dmi_data", _queueDepth);
    _setDgpsService = _nodeHandle.advertiseService("set_dgps",
      &PosLvNode::setDgps, this);
    _updater.setHardwareID("POS LV 220");
    _updater.add("TCP connection", this, &PosLvNode::diagnoseTCPConnection);
    _updater.add("System status", this, &PosLvNode::diagnoseSystemStatus);
    _vnsFreq = std::make_shared<diagnostic_updater::HeaderlessTopicDiagnostic>(
      "vehicle_navigation_solution", _updater,
      diagnostic_updater::FrequencyStatusParam(&_vnsMinFreq, &_vnsMaxFreq,
      0.1, 10));
    _vnpFreq = std::make_shared<diagnostic_updater::HeaderlessTopicDiagnostic>(
      "vehicle_navigation_performance", _updater,
      diagnostic_updater::FrequencyStatusParam(&_vnpMinFreq, &_vnpMaxFreq,
      0.1, 10));
    _dmiFreq = std::make_shared<diagnostic_updater::HeaderlessTopicDiagnostic>(
      "time_tagged_dmi_data", _updater,
      diagnostic_updater::FrequencyStatusParam(&_dmiMinFreq, &_dmiMaxFreq,
      0.1, 10));
    _updater.force_update();
  }

  PosLvNode::~PosLvNode() {
  }

/******************************************************************************/
/* Methods                                                                    */
/******************************************************************************/

  bool PosLvNode::setDgps(poslv::SetDGPS::Request& request,
      poslv::SetDGPS::Response& response) {
    static uint16_t transactionNumber = 0;
    if (request.mode != "cmr" && request.mode != "rtcm1" &&
        request.mode != "rtcm2") {
      response.response = false;
      response.message = "Unkown mode";
    }
    else {
      TCPConnectionClient tcpConnection(_deviceIpStr, _deviceControlPort);
      POSLVComTCP device(tcpConnection);
      std::shared_ptr<Packet> ctrlPacket(
        Factory<uint16_t, Message>::getInstance().create(90));
      ProgramControl& ctrlMsg =
        ctrlPacket->messageCast().typeCast<ProgramControl>();
      ctrlMsg.mControl = 0;
      ctrlMsg.mTransactionNumber = transactionNumber++;
      std::shared_ptr<Packet> dgpsPacket(
        Factory<uint16_t, Message>::getInstance().create(37));
      BaseGPS1Setup& dgpsMsg =
        dgpsPacket->messageCast().typeCast<BaseGPS1Setup>();
      dgpsMsg.mTransactionNumber = transactionNumber++;
      if (request.mode == "cmr")
        dgpsMsg.mBaseGPSInputType = 3;
      else if (request.mode == "rtcm1")
        dgpsMsg.mBaseGPSInputType = 1;
      else
        dgpsMsg.mBaseGPSInputType = 2;
      dgpsMsg.mLineControl = 0;
      dgpsMsg.mModemControl = 0;
      dgpsMsg.mConnectionControl = 0;
      dgpsMsg.mTimeoutLength = 0;
      try {
        Timer timer;
        device.writePacket(ctrlPacket);
        timer.sleep(0.1);
        device.writePacket(dgpsPacket);
        timer.sleep(0.1);
        ctrlMsg.mControl = 1;
        ctrlMsg.mTransactionNumber = transactionNumber++;
        device.writePacket(ctrlPacket);
        response.response = true;
        response.message = "Success";
      }
      catch (const IOException& e) {
        ROS_WARN_STREAM("IOException: " << e.what());
        response.response = false;
        response.message = e.what();
      }
      catch (const SystemException& e) {
        ROS_WARN_STREAM("SystemException: " << e.what());
        response.response = false;
        response.message = e.what();
      }
    }
    return true;
  }

  void PosLvNode::publishVehicleNavigationSolution(const ros::Time& timestamp,
      const VehicleNavigationSolution& vns) {
    if (_vehicleNavigationSolutionPublisher.getNumSubscribers() > 0) {
      auto vnsMsg = boost::make_shared<poslv::VehicleNavigationSolutionMsg>();
      vnsMsg->header.stamp = timestamp;
      vnsMsg->header.frame_id = _frameId;
      vnsMsg->header.seq = _vnsPacketCounter++;
      vnsMsg->timeDistance.time1 = vns.mTimeDistance.mTime1;
      vnsMsg->timeDistance.time2 = vns.mTimeDistance.mTime2;
      vnsMsg->timeDistance.distanceTag = vns.mTimeDistance.mDistanceTag;
      vnsMsg->timeDistance.timeType = vns.mTimeDistance.mTimeType;
      vnsMsg->timeDistance.distanceType = vns.mTimeDistance.mDistanceType;
      vnsMsg->latitude = vns.mLatitude;
      vnsMsg->longitude = vns.mLongitude;
      vnsMsg->altitude = vns.mAltitude;
      vnsMsg->northVelocity = vns.mNorthVelocity;
      vnsMsg->eastVelocity = vns.mEastVelocity;
      vnsMsg->downVelocity = vns.mDownVelocity;
      vnsMsg->roll = vns.mRoll;
      vnsMsg->pitch = vns.mPitch;
      vnsMsg->heading = vns.mHeading;
      vnsMsg->wanderAngle = vns.mWanderAngle;
      vnsMsg->trackAngle = vns.mTrackAngle;
      vnsMsg->speed = vns.mSpeed;
      vnsMsg->angularRateLong = vns.mAngularRateLong;
      vnsMsg->angularRateTrans = vns.mAngularRateTrans;
      vnsMsg->angularRateDown = vns.mAngularRateDown;
      vnsMsg->accLong = vns.mAccLong;
      vnsMsg->accTrans = vns.mAccTrans;
      vnsMsg->accDown = vns.mAccDown;
      vnsMsg->alignementStatus = vns.mAlignementStatus;
      _vehicleNavigationSolutionPublisher.publish(vnsMsg);
    }
    _vnsFreq->tick();
  }

  void PosLvNode::publishVehicleNavigationPerformance(
      const ros::Time& timestamp, const VehicleNavigationPerformance& vnp) {
    if (_vehicleNavigationPerformancePublisher.getNumSubscribers() > 0) {
      auto vnpMsg =
        boost::make_shared<poslv::VehicleNavigationPerformanceMsg>();
      vnpMsg->header.stamp = timestamp;
      vnpMsg->header.frame_id = _frameId;
      vnpMsg->header.seq = _vnpPacketCounter++;
      vnpMsg->timeDistance.time1 = vnp.mTimeDistance.mTime1;
      vnpMsg->timeDistance.time2 = vnp.mTimeDistance.mTime2;
      vnpMsg->timeDistance.distanceTag = vnp.mTimeDistance.mDistanceTag;
      vnpMsg->timeDistance.timeType = vnp.mTimeDistance.mTimeType;
      vnpMsg->timeDistance.distanceType = vnp.mTimeDistance.mDistanceType;
      vnpMsg->northPositionRMSError = vnp.mNorthPositionRMSError;
      vnpMsg->eastPositionRMSError = vnp.mEastPositionRMSError;
      vnpMsg->downPositionRMSError = vnp.mDownPositionRMSError;
      vnpMsg->northVelocityRMSError = vnp.mNorthVelocityRMSError;
      vnpMsg->eastVelocityRMSError = vnp.mEastVelocityRMSError;
      vnpMsg->downVelocityRMSError = vnp.mDownVelocityRMSError;
      vnpMsg->rollRMSError = vnp.mRollRMSError;
      vnpMsg->pitchRMSError = vnp.mPitchRMSError;
      vnpMsg->headingRMSError = vnp.mHeadingRMSError;
      vnpMsg->errorEllipsoidSemiMajor = vnp.mErrorEllipsoidSemiMajor;
      vnpMsg->errorEllipsoidSemiMinor = vnp.mErrorEllipsoidSemiMinor;
      vnpMsg->errorEllipsoidOrientation = vnp.mErrorEllipsoidOrientation;
      _vehicleNavigationPerformancePublisher.publish(vnpMsg);
    }
    _vnpFreq->tick();
  }

  void PosLvNode::publishTimeTaggedDMIData(
      const ros::Time& timestamp, const TimeTaggedDMIData& dmi) {
    if (_timeTaggedDMIDataPublisher.getNumSubscribers() > 0) {
      auto dmiMsg = boost::make_shared<poslv::TimeTaggedDMIDataMsg>();
      dmiMsg->header.stamp = timestamp;
      dmiMsg->header.frame_id = _frameId;
      dmiMsg->header.seq = _dmiPacketCounter++;
      dmiMsg->timeDistance.time1 = dmi.mTimeDistance.mTime1;
      dmiMsg->timeDistance.time2 = dmi.mTimeDistance.mTime2;
      dmiMsg->timeDistance.distanceTag = dmi.mTimeDistance.mDistanceTag;
      dmiMsg->timeDistance.timeType = dmi.mTimeDistance.mTimeType;
      dmiMsg->timeDistance.distanceType = dmi.mTimeDistance.mDistanceType;
      dmiMsg->signedDistanceTraveled = dmi.mSignedDistanceTraveled;
      dmiMsg->unsignedDistanceTraveled = dmi.mUnsignedDistanceTraveled;
      dmiMsg->dmiScaleFactor = dmi.mDMIScaleFactor;
      dmiMsg->dataStatus = dmi.mDataStatus;
      dmiMsg->dmiType = dmi.mDMIType;
      dmiMsg->dmiDataRate = dmi.mDMIDataRate;
      _timeTaggedDMIDataPublisher.publish(dmiMsg);
    }
    _dmiFreq->tick();
  }

  void PosLvNode::diagnoseTCPConnection(
      diagnostic_updater::DiagnosticStatusWrapper& status) {
    if (_tcpConnection && _tcpConnection->isOpen()) {
      if (_lastInterVnsTime)
        status.add("Inter VNS packet time [s]", _lastInterVnsTime);
      if (_lastInterVnpTime)
        status.add("Inter VNP packet time [s]", _lastInterVnpTime);
      if (_lastInterDmiTime)
        status.add("Inter DMI packet time [s]", _lastInterDmiTime);
      status.summaryf(diagnostic_msgs::DiagnosticStatus::OK,
        "TCP connection opened on %s:%d.",
        _tcpConnection->getServerIP().c_str(),
        _tcpConnection->getPort());
    }
    else
     status.summaryf(diagnostic_msgs::DiagnosticStatus::ERROR,
      "TCP connection closed on %s:%d.", _deviceIpStr.c_str(), _devicePort);
  }

  void PosLvNode::diagnoseSystemStatus(
      diagnostic_updater::DiagnosticStatusWrapper& status) {
    if (_alignStatusMsgs.count(_alignStatus))
      status.add("Alignment status", _alignStatusMsgs[_alignStatus]);
    if (_gpsStatusMsgs.count(_navStatus1))
      status.add("Primary GPS status", _gpsStatusMsgs[_navStatus1]);
    if (_gpsStatusMsgs.count(_navStatus2))
      status.add("Secondary GPS status", _gpsStatusMsgs[_navStatus2]);
    if (_gamsStatusMsgs.count(_gamsStatus))
      status.add("GAMS solution status", _gamsStatusMsgs[_gamsStatus]);
    if (_iinStatusMsgs.count(_iinStatus))
      status.add("IIN processing status", _iinStatusMsgs[_iinStatus]);
    status.add("RTCM 1 count", _rtcm1Count);
    status.add("RTCM 3 count", _rtcm3Count);
    status.add("RTCM 9 count", _rtcm9Count);
    status.add("RTCM 18 count", _rtcm18Count);
    status.add("RTCM 19 count", _rtcm19Count);
    status.add("CMR 0 count", _cmr0Count);
    status.add("CMR 1 count", _cmr1Count);
    status.add("CMR 2 count", _cmr2Count);
    status.add("CMR 94 count", _cmr94Count);
    std::bitset<32> statusA(_generalStatusA);
    if (statusA.test(7))
      status.summaryf(diagnostic_msgs::DiagnosticStatus::OK,
        "Full navigation solution");
    else
      status.summaryf(diagnostic_msgs::DiagnosticStatus::WARN,
        "Incomplete navigation solution");
  }

  void PosLvNode::spin() {
    _tcpConnection = std::make_shared<TCPConnectionClient>(_deviceIpStr,
      _devicePort);
    POSLVComTCP device(*_tcpConnection);
    Timer timer;
    while (_nodeHandle.ok()) {
      try {
        std::shared_ptr<Packet> packet = device.readPacket();
        const ros::Time timestamp = ros::Time::now();
        if (packet->instanceOfGroup()) {
          const Group& group = packet->groupCast();
          if (group.instanceOf<VehicleNavigationSolution>()) {
            const VehicleNavigationSolution& vns =
              group.typeCast<VehicleNavigationSolution>();
            publishVehicleNavigationSolution(timestamp, vns);
            if (_lastVnsTimestamp)
              _lastInterVnsTime = vns.mTimeDistance.mTime2 - _lastVnsTimestamp;
            _lastVnsTimestamp = vns.mTimeDistance.mTime2;
            _alignStatus = vns.mAlignementStatus;
          }
          else if (group.instanceOf<VehicleNavigationPerformance>()) {
            const VehicleNavigationPerformance& vnp =
              group.typeCast<VehicleNavigationPerformance>();
            publishVehicleNavigationPerformance(timestamp, vnp);
            if (_lastVnpTimestamp)
              _lastInterVnpTime = vnp.mTimeDistance.mTime2 - _lastVnpTimestamp;
            _lastVnpTimestamp = vnp.mTimeDistance.mTime2;
          }
          else if (group.instanceOf<TimeTaggedDMIData>()) {
            const TimeTaggedDMIData& dmi = group.typeCast<TimeTaggedDMIData>();
            publishTimeTaggedDMIData(timestamp, dmi);
            if (_lastDmiTimestamp)
              _lastInterDmiTime = dmi.mTimeDistance.mTime2 - _lastDmiTimestamp;
            _lastDmiTimestamp = dmi.mTimeDistance.mTime2;
          }
          else if (group.instanceOf<PrimaryGPSStatus>()) {
            const PrimaryGPSStatus& gps = group.typeCast<PrimaryGPSStatus>();
            _navStatus1 = gps.mNavigationSolutionStatus;
          }
          else if (group.instanceOf<SecondaryGPSStatus>()) {
            const SecondaryGPSStatus& gps =
              group.typeCast<SecondaryGPSStatus>();
            _navStatus2 = gps.mNavigationSolutionStatus;
          }
          else if (group.instanceOf<GAMSSolutionStatus>()) {
            const GAMSSolutionStatus& gams =
              group.typeCast<GAMSSolutionStatus>();
            _gamsStatus = gams.mSolutionStatus;
          }
          else if (group.instanceOf<IINSolutionStatus>()) {
            const IINSolutionStatus& iin =
              group.typeCast<IINSolutionStatus>();
            _iinStatus = iin.mIINProcessingStatus;
          }
          else if (group.instanceOf<GeneralStatusFDIR>()) {
            const GeneralStatusFDIR& stat =
              group.typeCast<GeneralStatusFDIR>();
            _generalStatusA = stat.mGeneralStatusA;
            _generalStatusB = stat.mGeneralStatusB;
            _generalStatusC = stat.mGeneralStatusC;
            _fdirLevel1Status = stat.mFDIRLevel1Status;
            _fdirLevel2Status = stat.mFDIRLevel2Status;
            _fdirLevel4Status = stat.mFDIRLevel4Status;
            _fdirLevel5Status = stat.mFDIRLevel5Status;
            std::bitset<32> statusC(_generalStatusC);
            if (statusC.test(18))
              _rtcm1Count++;
            if (statusC.test(19))
              _rtcm3Count++;
            if (statusC.test(20))
              _rtcm9Count++;
            if (statusC.test(21))
              _rtcm18Count++;
            if (statusC.test(22))
              _rtcm19Count++;
            if (statusC.test(23))
              _cmr0Count++;
            if (statusC.test(24))
              _cmr1Count++;
            if (statusC.test(25))
              _cmr2Count++;
            if (statusC.test(26))
              _cmr94Count++;
          }
        }
      }
      catch (const IOException& e) {
        ROS_WARN_STREAM("IOException: " << e.what());
        ROS_WARN_STREAM("Retrying in " << _retryTimeout << " [s]");
        timer.sleep(_retryTimeout);
      }
      catch (const SystemException& e) {
        ROS_WARN_STREAM("SystemException: " << e.what());
        ROS_WARN_STREAM("Retrying in " << _retryTimeout << " [s]");
        timer.sleep(_retryTimeout);
      }
      catch (const TypeCreationException<unsigned short>& e) {
      }
      _updater.update();
      ros::spinOnce();
    }
  }

  void PosLvNode::getParameters() {
    _nodeHandle.param<std::string>("ros/frame_id", _frameId,
      "/poslv_link");
    _nodeHandle.param<int>("ros/queue_depth", _queueDepth, 100);
    _nodeHandle.param<std::string>("connection/device_ip", _deviceIpStr,
      "129.132.39.171");
    _nodeHandle.param<int>("connection/device_port", _devicePort, 5602);
    _nodeHandle.param<int>("connection/device_control_port", _deviceControlPort,
      5601);
    _nodeHandle.param<double>("connection/retry_timeout", _retryTimeout, 1);
    _nodeHandle.param<double>("diagnostics/vns_min_freq", _vnsMinFreq, 80);
    _nodeHandle.param<double>("diagnostics/vns_max_freq", _vnsMaxFreq, 120);
    _nodeHandle.param<double>("diagnostics/vnp_min_freq", _vnpMinFreq, 0.8);
    _nodeHandle.param<double>("diagnostics/vnp_max_freq", _vnpMaxFreq, 1.2);
    _nodeHandle.param<double>("diagnostics/dmi_min_freq", _dmiMinFreq, 80);
    _nodeHandle.param<double>("diagnostics/dmi_max_freq", _dmiMaxFreq, 120);
  }

}
